File: wordlist_combiner.py
# This script combines large files into a "master" file using hyperthreading and chunking to not crash with large files. 
# Usage: python3 wordlist_combiner.py <wordlist1> <wordlist2>
import argparse
from concurrent.futures import ThreadPoolExecutor

CHUNK_SIZE = 32768  # Adjust the chunk size based on your needs/RAM availability

def process_chunk(chunk):
    return set(chunk.splitlines())

def process_file(file):
    with open(file, "r", encoding="utf-8") as f:
        chunks = iter(lambda: f.read(CHUNK_SIZE), '')
        return set(entry for chunk in chunks for entry in process_chunk(chunk))

def combine_text_files(files):
    master_list = set()

    with ThreadPoolExecutor() as executor:
        file_contents = executor.map(process_file, files)
        master_list.update(entry for file_set in file_contents for entry in file_set)

    # Write the deduplicated entries to the master file
    with open("master.txt", "w", encoding="utf-8") as master_file:
        for entry in master_list:
            master_file.write(entry + '\n')

    print("Combining and deduplicating completed. Master list saved to master.txt.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Combine and deduplicate entries from text files into a master list.")
    parser.add_argument("files", nargs="+", help="List of text files to process")
    args = parser.parse_args()

    combine_text_files(args.files)

File: wordlist_combiner.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};
use std::collections::HashSet;
use std::sync::Mutex;
use rayon::prelude::*;

fn process_chunk(chunk: String) -> HashSet<String> {
    chunk.lines().map(String::from).collect()
}

fn process_file(file: &str) -> io::Result<HashSet<String>> {
    let file = File::open(file)?;
    let reader = BufReader::new(file);
    let mut set = HashSet::new();

    for chunk in reader.lines().map(|line| line.unwrap()) {
        set.extend(process_chunk(chunk));
    }

    Ok(set)
}

fn combine_text_files(files: Vec<String>) -> io::Result<()> {
    let master_list = Mutex::new(HashSet::new());

    files.par_iter().map(|file| {
        process_file(file).map(|file_set| {
            let mut master = master_list.lock().unwrap();
            master.extend(file_set);
        })
    }).collect::<Result<Vec<_>, _>>()?;

    let mut master_file = File::create("master.txt")?;
    let master = master_list.lock().unwrap();

    for entry in master.iter() {
        writeln!(master_file, "{}", entry)?;
    }

    println!("Combining and deduplicating completed. Master list saved to master.txt.");
    Ok(())
}

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();
    if args.is_empty() {
        eprintln!("Usage: program <file1> <file2> ...");
        std::process::exit(1);
    }

    if let Err(e) = combine_text_files(args) {
        eprintln!("Error: {}", e);
        std::process::exit(1);
    }
}
